fix apiserver to advertise IPv6 endpoints if bound to IPv6

danwinship opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
If you run `kube-apiserver --bind-address ::` on a machine with both an IPv4 address and an IPv6 address, it will bind to all interfaces and then advertise the IPv4 address in the `kubernetes.default` Endpoints, exactly like it would have if you had said `kube-apiserver --bind-address 127.0.0.1`. You can't get it to advertise its IPv6 address without explicitly passing `--advertise-address`, and in that case you have to figure it out yourself (which in turn means that you need a host-specific config file).

This fixes `utilnet.ChooseBindAddress()` to respect the address family of the passed-in address and try to return another address of the same family. Within k/k, `ChooseBindAddress` is only used by kube-apiserver (and by kubeadm for creating the kube-apiserver config). I'm not sure if things in other repos use that function and depend on the current blatantly-IPv4-biased behavior.

Beyond the `ChooseBindAddress` fix, this also fixes another apiserver-specific bug, which is that if you didn't explicitly specify an advertise address, it would always autodetect it based on the insecure bind address rather than the secure bind address, because the secure config code wasn't checking if `s.ExternalAddress` was nil (and `IsUnspecified()` returns `false` for a nil IP, so `SecureServingOptions.DefaultExternalAddress()` would end up returning nil, so it would fall back to figuring out the advertise address from the insecure options instead.)

**Which issue(s) this PR fixes**:
Fixes #72404

**Special notes for your reviewer**:
I tried to minimize the changes to `k8s.io/apimachinery/pkg/util/net/interface.go` rather than rewriting it in the nicest possible way...

**Does this PR introduce a user-facing change?**:
```release-note
If given an IPv6 bind-address, kube-apiserver will now advertise an IPv6 endpoint for the kubernetes.default service.
```

/sig network
/area ipv6
cc @khenidak @aojea (not sure who else might have opinions here?)

danwinship committed:
Fix apiserver to advertise IPv6 endpoints if bound to IPv6

Also rename utilnet.ChooseBindAddress() to ResolveBindAddress(), to
better describe its functionality.

danwinship commented:
so this is failing because I forgot to update the unit tests, but before I do that I'd like feedback on whether the idea makes sense...

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:
```diff
@@ -406,18 +424,19 @@ func chooseHostInterfaceFromRoute(routes []Route, nw networkInterfacer) (net.IP,
+// ChooseBindAddress returns an IP address for a daemon, based on the given bindAddress:
+// If bindAddress is unset, it returns the host's default IP, as with ChooseHostInterface().
+// If bindAddress is unspecified or loopback, it returns the default IP of the same
+// address family as bindAddress.
```
> it returns the default IP of the same address family as bindAddress.

does this match the stdlib bind behavior? if I bind to `0.0.0.0`, do I only get bound to all ipv4 interfaces?

also, don't we use 0.0.0.0 as a stand-in for "all" in some of our default options? what would this change do to defaults on an ipv6-only machine?

danwinship replied:
> does this match the stdlib bind behavior? if I bind to 0.0.0.0, do I only get bound to all ipv4 interfaces?

The PR doesn't change what address the apiserver actually binds to. The function name is misleading. (The apiserver always binds to the address you passed in as `--bind-address`. `ChooseBindAddress()` is used to figure out the default value for `--advertise-address`.) Probably we should rename the function...

> also, don't we use "0.0.0.0" as a stand-in for "all" in some of our default options? what would this change do to defaults on an ipv6-only machine?

Ah, yes, that would break with this PR. I guess for compatibility it should *prefer* the passed-in address family, but not require it.

neolit123 replied:
> I guess for compatibility it should prefer the passed-in address family, but not require it.

isn't this already handled by the logic on line 420-421?

aojea commented:
Fixes https://github.com/kubernetes/kubernetes/issues/72404

danwinship commented:
Fixed it to prefer-but-not-require IPv6, fixed (and extended) unit tests, renamed `ChooseBindAddress` to `ResolveBindAddress` which I think describes it better without being too verbose.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface_test.go:
```diff
@@ -575,24 +582,29 @@ func TestGetIPFromHostInterfaces(t *testing.T) {
-		{"no matches", networkInterfaceWithOnlyLinkLocals{}, nil, "no acceptable"},
-		{"ipv4", validNetworkInterface{}, net.ParseIP("10.254.71.145"), ""},
-		{"ipv6", ipv6NetworkInterface{}, net.ParseIP("2001::200"), ""},
+		{"fail get I/Fs", failGettingNetworkInterface{}, preferV4, nil, "failed getting all interfaces"},
```
should we add a test case for `0.0.0.0` and `::` addresses? 
it will cover liggitt 's question for the future

aojea replied:
IC, I was asking because looking at the current test cases I saw one test case with link locals
`networkInterfaceWithOnlyLinkLocals` despite it's discarded in the `ResolveBindAddress` too.

neolit123 commented:
/approve
for kubeadm

caesarxuchao commented:
/assign @deads2k

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6

aojea commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6
unrelated failures, most probably flaky
/lgtm

deads2k commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface_test.go:
```diff
@@ -128,6 +128,9 @@ var (
 	routeV6    = []Route{ipv6Route}
 	bothRoutes = []Route{ipv4Route, ipv6Route}
+
+	preferV4 = []AddressFamily{familyIPv4, familyIPv6}
```
move this to interface.go and use it instead of the hardcoded slices to improve readability.

deads2k commented:
minor comment, lgtm otherwise

/approve
/hold for minor readability update.  feel free to self-unhold after.

danwinship commented:
Moved the "prefer" definitions, and added a `type AddressFamilyPreference []AddressFamily` for further self-documentingness.
/hold cancel

danwinship commented:
/retest

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6

aojea commented:
ipv6 failures are unrelated to this PR
sgtm, defer the lgtm to @deads2k

neolit123 commented:
/retest

@aojea LGTM is on you! :)

aojea commented:
/lgtm
:) thx neolit
